Minor dynlink cleanups

shindere opened the pull request:
This PR contains two independent commits. The first one stops configuring
which warnings are enabled / disabled in several dynlink source files, as
this is either redundent with what is done in the build system or not
required (warning 30).

The second commit stops relying on Misc.fatal_error and rather calls
`failwith`, as a small step towards getting rid of the dependencies of
Dynlink on the compiler internals, in coneection with the emancipatin work
done in #11996.

shindere committed:
Dynlink: release the dependency on Misc.fatal_error

stedolan commented:
The change looks fine, but I think you need to promote some testsuite outputs because line numbers changed.

gasche commented:
`backtrace_dynlink` is still failing, the `.flambda.reference` output should also be updated

shindere committed:
Dynlink: stop configuring warnings in source files

Most of the settings in these files were redundent with those of the build
system anyway. The only warning which was disabled in the source
files but not in the build system was warning 30 (duplicate-definitions)
but since the current implementations do not trigger it anyway it
seems better to not disable it, at least as long as there is no
need to do so.

shindere commented:
Gabriel Scherer (2023/09/25 05:10 -0700):
> `backtrace_dynlink` is still failing, the `.flambda.reference` output
> should also be updated

Yep, fixed now. Sorry I missed that one. Thanks!

shindere commented:
Stephen Dolan (2023/09/25 04:32 -0700):
> The change looks fine, but I think you need to promote some testsuite
> outputs because line numbers changed.

Yes, done, thanks. Sorry I overlooked those. Will be more careful next
time, hopefully.

gasche commented:
Note: I wonder if the per-file warning attributes could be related to the `dune` build system for the compiler. (It might be that the change breaks some parts of the `dune` build.)

shindere commented:
Gabriel Scherer (2023/09/25 06:26 -0700):
> Note: I wonder if the per-file warning attributes could be related to
> the `dune` build system for the compiler. (It might be that the change
> breaks some parts of the `dune` build.)

Well, these files are not really special in any way, so if Dune is able
to compile the other source files it should be able to compile these,
too. The patch really removes a build irregularity, IMO, so I
don't really see how it could break anything.
